src/lib/registration/widgets/registration_code.rb

```diff
@@ -0,0 +1,166 @@
+# encoding: utf-8
+
+# ------------------------------------------------------------------------------
+# Copyright (c) 2017 SUSE LLC
+#
+#
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of version 2 of the GNU General Public License as published by the
+# Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, contact SUSE.
+#
+# To contact SUSE about this file by physical or electronic mail, you may find
+# current contact information at www.suse.com.
+# ------------------------------------------------------------------------------
+
+require "yast"
+require "cwm/widget"
+require "uri"
+require "registration/registration_ui"
+require "registration/sw_mgmt"
+
+module Registration
+  module Widgets
+    class RegistrationCode < CWM::InputField
+      VALID_URL_SCHEMES = ["http", "https"].freeze
+
+      def initialize
+        textdomain "registration"
+      end
+
+      def label
+        _("Registration Code or SMT Server URL")
+      end
+
+      # Initialize the widget with stored values
+      def init
+        reg_code = options.reg_code.to_s
+
+        self.value = reg_code.empty? ? (options.custom_url || boot_url) : options.reg_code
+      end
+
+      # Set registration options according to the value
+      def store
+        if valid_url?
+          options.reg_code   = ""
+          options.custom_url = value
+        else
+          options.reg_code   = value
+          options.custom_url = default_url
+        end
+
+        register
```

Comment:
if register report false, we ignore it. Does it show popup itself, or it should be showed here?